src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb

```diff
@@ -493,17 +592,29 @@ def create_network_interfaces(resource_group_name, vm_name, location, vm_props,
           nic_params[:load_balancer] = nil
           nic_params[:application_gateway] = nil
         end
-        @azure_client.create_network_interface(resource_group_name, nic_params)
-        network_interfaces.push(@azure_client.get_network_interface_by_name(resource_group_name, nic_name))
+        tasks_creating.push(
+          Concurrent::Future.execute do
+            @azure_client.create_network_interface(resource_group_name, nic_params)
+            @azure_client.get_network_interface_by_name(resource_group_name, nic_name)
+          end
+        )
       end
-      network_interfaces
+      tasks_creating.map(&:wait)
+      network_interfaces = tasks_creating.map(&:value!)
     end
 
     def delete_possible_network_interfaces(resource_group_name, vm_name)
       network_interfaces = @azure_client.list_network_interfaces_by_keyword(resource_group_name, vm_name)
+      tasks = []
       network_interfaces.each do |network_interface|
-        @azure_client.delete_network_interface(resource_group_name, network_interface[:name])
+        tasks.push(
+          Concurrent::Future.execute do
+            @azure_client.delete_network_interface(resource_group_name, network_interface[:name])
+          end
+        )
       end
+      tasks.map(&:wait)
```

Comment:
not familiar with the concurrency lib.
but it looks strange to call wait and wait! at the same place.

Reply:
If only `.wait!` is used, then one task raises an error and abort the CPI, while some other tasks aren't completed. It means some resources may be in provisioning. Is my understanding correct?

And, it's better to add more details to the commend before `.wait`. For example, adding `The purpose of calling .wait before .wait! is to make sure that all processes are completed before raising an error. It is important when creating a VM, because on Azure you can't delete a resource that is on-provisioning.`

Reply:
yes, correct. will add the comment.